### DISCUSSION TITLE
Input Forms

### DISCUSSION THREAD

**Discussion by wwwat** [telegraf/telegraf#1246] [Ideas]
I can see many cases, where interaction with a bot can be thought of as filling up an input form: you go step by step, providing the bot with answers to given questions, in the end you might want to check/fix one of the input field values and confirm. This sounds similar to the way wizard scene is functioning, however, in wizard scene every step is dependent on the previous one (e.g. you have to reply with a question in step 1, and process user input in step 2).

Maybe there is an existing solution for this, however, I wasn't able to find it. If that's the case, pointing in the direction of such tool will be highly appreciated.
If not, does it make sense to organize it as part of this project or as an outside library for it?

**Comment by AleksandrZhukov** [comment]
Faced with the same inconvenience, made a small api proposal
<img width="1792" alt="Screenshot 2021-01-21 at 01 16 31" src="https://user-images.githubusercontent.com/8876116/105340397-7f394f00-5be6-11eb-90ce-cc793a701d58.png">

**Reply by wojpawlik** [reply]
`Composer` doesn't know about `WizardScene` and it should stay this way. However, `BaseScene` extends `Composer` and has `enter` method and `enterHandler` property, maybe that could be used?

**Reply by AleksandrZhukov** [reply]
Yes sure, that's just a draft, it can be some new entity, eg `WizardStep` that will extend `Composer`.

> BaseScene extends Composer and has enter method

Yes, I know about that, but it can be used only for entering the wizard, but not for entering each step. correct me, if I'm wrong.

**Reply by wojpawlik** [reply]
Yes. I mean, perhaps `WizardScene` can be altered to be able to take advantage of `BaseScene` steps, @KnorpelSenf.

**Reply by KnorpelSenf** [reply]
Not quite sure what you're missing.

Scenes: based on names of the states, allows arbitrary jumping between states only.
Wizards: based on indices in an array of the states, allows arbitrary jumping between states, allows going back and forth in the array.

Enter/leave middleware allows to “ask questions” when transitioning between states.

I feel like that should cover it, but I did not implement many different forms yet, so perhaps I'm missing something. Could you maybe give an example of a form, how you'd implement with wizards (if possible), and how a new API is more convenient?

**Comment by wojpawlik** [comment]
Definitely make it a separate library. Don't base it on `WizardScene`, at this point it's so messy nobody fully understands it. `ctx.session` is prone to race conditions. See https://www.npmjs.com/package/telegraf-stateless-question.
